src/methods.ts

```diff
@@ -254,6 +347,11 @@ export class Method extends AbstractMethodField {
    */
   private code: any;
 
+  private numOpcodeExecs = 0;
+  private jitThreshold = 0;
+  private compiledFunctions: Function[] = [];
```

Comment:
> My version of re-JIT emits the function source again.

Gotcha; that's fine with me. It reduces memory usage, too.

> I do have a branch that re-JITs whenever byte code is updated to one of the "fast" opcodes. However, I didn't observe significant performance improvement with it.

In this case, it's all up to the heuristic you are using. If the method *interprets* a significant number of basic blocks after JITting, then it is a good candidate for re-JITting.

One possible issue here is that if there are particular blocks that cannot be JITted due to blacklisted opcodes (e.g. `wide`). A hot function with one of these blacklisted opcodes may repeatedly try to be JITted and fail. In this case, we can have a `blacklist` boolean or something to prevent that from happening.

How does your `switch` statement handle jumps from interpreted to compiled basic blocks? I'm thinking we can make the `switch` smarter and have a case statement for each interpreted basic block; it could do something like:

```{js}
while (pc < endOfBB && pc >= startOfBB) {
  // interpret
}
```

Also, note that, unlike traditional basic blocks, DoppioJVM basic blocks end when there is a method call or a classloading event, since those are asynchronous. I believe classloading events are completely ruled out if the blocks you are compiling have been run once and have the `fast` versions of opcodes. It looks like you already use that same definition, since you close traces at function call boundaries.